src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php

```diff
@@ -221,6 +221,10 @@ public static function getCssToXPathTestData()
             ['div#container p', "div[@id = 'container']/descendant-or-self::*/p"],
             [':scope > div[dataimg="<testmessage>"]', "*[1]/div[@dataimg = '<testmessage>']"],
             [':scope', '*[1]'],
+            ['e:has(> f)', 'e[./f]'],
+            ['e:has(~ f)', 'e[following-sibling::f]'],
+            ['e:has(+ f)', "e[following-sibling::*[(name() = 'f') and (position() = 1)]]"],
+            ['e:has(f)', 'e[descendant-or-self::f]'],
```

Comment:
It would be great to add some `:has()` selector to `getHtmlShakespearTestData` or `getHtmlIdsTestData` to also have tests ensuring we get the right elements

Reply:
Okay, I added relevant tests on both of these datasets. Anyway, I resolved here (https://github.com/symfony/symfony/pull/50729) the issue linked with `testHtmlIds` tests. Please tell me if I have to bring here also changes made there.